Add guide for subclassing an existing provider (#492)
ManageIQ/guides: providers/subclassing_an_existing_provider.md

      create  Gemfile
      create  LICENSE.txt
      create  Rakefile
      create  README.md
      create  bin/ci/after_script
      create  bin/rails
      create  bin/setup
      create  bin/update
      create  bundler.d
      create  bundler.d/.keep
      create  config/secrets.defaults.yml
      create  config/settings.yml
      create  lib/manageiq-providers-awesome_private_cloud.rb
      create  lib/manageiq/providers/awesome_private_cloud/engine.rb
      create  lib/manageiq/providers/awesome_private_cloud/version.rb
      create  lib/tasks/README.md
      create  lib/tasks_private/spec.rake
      create  locale
      create  locale/.keep
      create  spec/factories
      create  spec/support
      create  spec/spec_helper.rb
      insert  /home/grare/adam/src/manageiq/manageiq/Gemfile
      create  spec/models/manageiq/providers/awesome_private_cloud
        gsub  lib/tasks_private/spec.rake
      insert  .yamllint
      append  spec/spec_helper.rb

```

Follow the steps from [Creating a new provider](writing_a_new_provider.md) for setting up your new plugin for local development by adding to your `bundler.d/override.rb` and symlinking your `spec/manageiq` directory.

> You'll notice however that we chose "--no-scaffolding" option, this is going to allow us to define the classes that we need as subclasses of the OpenStack provider.

Reviewer: This makes me wonder if we should have a `--subclasses ManageIQ::Providers::AwesomeCloud`, which, if --scaffolding is still set, will automatically scaffold the subclasses

Author: Yes that is the plan but until then wanted to document how to currently do it

Reviewer: Oh yeah definitely

### Creating the subclasses

Now that we have an empty plugin we have to start adding the subclasses that we need for the provider to work.

First let's create the initial directories that will hold our subclasses:
```bash
$ mkdir -p app/models/manageiq/providers/awesome_private_cloud/{cloud_manager,network_manager,storage_manager} app/models/manageiq/providers/awesome_private_cloud/inventory/{collector,parser,persister}
```

> ```ruby

Reviewer: I think there's a sentence missing before this; the previous comment talks about directories.

Author: https://github.com/ManageIQ/guides/pull/492/files#diff-f19823aaa72a5e378b421bb97147c5ac793580e22c74efa57d92f72db211d581R72
got ahead of myself talking about sti

Reviewer: I think it got lost somehow, and I don't see it in the diff?

Author: Yeah switched machines and didn't pull latest from this branch, fixed

This is where `ActsAsStiLeafClass` comes in.  It overrides the `type_condition` method which typically does `sti_names  = ([self] + descendants).map(&:sti_name)` to get a list of types, and replaces it with just `[sti_name]`.

### Creating the subclasses

After that little detour we can get back to creating our provider.  The summary here is that any class which is a subclass of ActiveRecord::Base must include `Class.include(ActsAsStiLeafClass)`.

So if we go back to our base provider (OpenStack) we need to find each `ActiveRecord::Base` subclass and create a corresponding subclass in our provider.

Go through the OpenStack cloud_manager/* and for each `ActiveRecord::Base` subclass create a file like the following:
```ruby
ManageIQ::Providers::Openstack::CloudManager::AuthKeyPair.include(ActsAsStiLeafClass)

class ManageIQ::Providers::AwesomePrivateCloud::CloudManager::AuthKeyPair < ManageIQ::Providers::Openstack::CloudManager::AuthKeyPair
end
```

> And add a `require_nested :AuthKeyPair` in `ManageIQ::Providers::AwesomePrivateCloud::CloudManager`.

Reviewer: ```suggestion
followed by a `require_nested :AuthKeyPair` in `ManageIQ::Providers::AwesomePrivateCloud::CloudManager`.
```

> Rinse&repeat until in this example you have:

Reviewer: ```suggestion
Rinse and repeat until, in this example, you have:
```

## Subclassing an existing provider

We have already covered creating a new provider from scratch in [Creating a new provider](writing_a_new_provider.md), however it is also common for a provider to use the API from another.  For example there are a number of Cloud Services which host Kubernetes and expose the standard k8s API with different authentication.  The Openstack API is also an industry standard for private clouds and is widely implemented by other providers.

In the case where the existing provider API is already a part of ManageIQ it is possible to "subclass" the existing provider plugin and only implement the differences or possibly just provide a more specific name and logo that users will recognize.

> For our example here we are going to add an on-premise version of our existing AwesomeCloud from [Creating a new provider](writing_a_new_provider.md), called AwesomePrivateCloud.  This is going to have an OpenStack compatible API but we are going to make some minor changes to the available options, taking advantage of the fact that we have a specific provider plugin.

Reviewer: ```suggestion
For our example here we are going to add an on-premise version of our existing AwesomeCloud from [Creating a new provider](writing_a_new_provider.md), called AwesomePrivateCloud.  This is going to have an AwesomeCloud compatible API but we are going to make some minor changes to the available options, taking advantage of the fact that we have a specific provider plugin.
```

      create  spec/support
      create  spec/spec_helper.rb
      insert  /home/grare/adam/src/manageiq/manageiq/Gemfile
      create  spec/models/manageiq/providers/awesome_private_cloud
        gsub  lib/tasks_private/spec.rake
      insert  .yamllint
      append  spec/spec_helper.rb

```

Follow the steps from [Creating a new provider](writing_a_new_provider.md) for setting up your new plugin for local development by adding to your `bundler.d/override.rb` and symlinking your `spec/manageiq` directory.

You'll notice however that we chose "--no-scaffolding" option, this is going to allow us to define the classes that we need as subclasses of the OpenStack provider.

### Creating the subclasses

Now that we have an empty plugin we have to start adding the subclasses that we need for the provider to work.

First let's create the initial directories that will hold our subclasses:
```bash
$ mkdir -p app/models/manageiq/providers/awesome_private_cloud/{cloud_manager,network_manager,storage_manager} app/models/manageiq/providers/awesome_private_cloud/inventory/{collector,parser,persister}
```

Now we can start defining our main manager class that will inherit from OpenStack:
```ruby

> ManageIQ::Providers::Openstack::CloudManager.include(ActsAsStiLeafClass)

Reviewer: ```suggestion
ManageIQ::Providers::AwesomeCloud::CloudManager.include(ActsAsStiLeafClass)
```

Author: OpenStack was intentional, the idea here was we had an on-premise provider which implements an OpenStack compatible API.

`AwesomeCloud` was our example of a public-cloud

Reviewer: Oh I see...I think what's confusing is it's named AwesomePrivateCloud, and name-wise I thought it was a subclass of AwesomeCloud because they are similarly named.

Author: Yeah in terms of back-story I was going for "If AwesomeCloud wanted to release an on-premise version"

> class ManageIQ::Providers::AwesomePrivateCloud::CloudManager < ManageIQ::Providers::Openstack::CloudManager

Reviewer: ```suggestion
class ManageIQ::Providers::AwesomePrivateCloud::CloudManager < ManageIQ::Providers::AwesomeCloud::CloudManager
```

```

This is exactly what we want in most cases, however if we're subclassing another provider such as `ManageIQ::Providers::Openstack::CloudManager::Vm`, if we do `ManageIQ::Providers::Openstack::CloudManager::Vm.all` we will accidentally retrieve all OpenStack VMs _and_ all subclassed provider VMs.

This is where `ActsAsStiLeafClass` comes in.  It overrides the `type_condition` method which typically does `sti_names  = ([self] + descendants).map(&:sti_name)` to get a list of types, and replaces it with just `[sti_name]`.

### Creating the subclasses

After that little detour we can get back to creating our provider.  The summary here is that any class which is a subclass of ActiveRecord::Base must include `Class.include(ActsAsStiLeafClass)`.

> So if we go back to our base provider (OpenStack) we need to find each `ActiveRecord::Base` subclass and create a corresponding subclass in our provider.

Reviewer: ```suggestion
So if we go back to our base provider (AwesomeCloud) we need to find each `ActiveRecord::Base` subclass and create a corresponding subclass in our provider.
```
